feat(GroupExtension/Abelian): define `OfMulDistribMulAction.equivH2`

yu-yama opened the pull request:
Mainly defines:

- `structure GroupExtension.OfMulDistribMulAction N G [MulDistribMulAction G N]`: group extensions of `G` by `N` where the multiplicative action of `G` on `N` is the conjugation
- `structure GroupExtension.OfMulDistribMulActionWithSection N G [MulDistribMulAction G N]`: group extensions with specific choices of sections
- `def GroupExtension.OfMulDistribMulAction.equivH2`: a bijection between the equivalence classes of group extensions and $H^2 (G, N)$

---

- [x] depends on: #20802 (split PR: contains changes to the `Defs` file)
- [x] depends on: #20998 (split PR: mainly adds the `Basic` file)
- [ ] depends on: #26670 (split PR: adds the first part of the `Abelian` file)

Here is a relevant TODO in Mathlib:
https://github.com/leanprover-community/mathlib4/blob/4e9fa40d7c480937e09cd6e47a591bd6f3b8be42/Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean#L46-L48

I would appreciate your comments.

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

yu-yama committed:
feat(GroupExtension/Abelian): define `GroupExtension.ofMulDistribMulAction.equivH2`

yu-yama committed:
docs: populate missing docstrings

erdOne commented:
Thanks for the contribution! 
This PR is quite large. Is it possible to split up the PR into smaller chunks? (e.g. the changes to three files could be three  sequential PRs)
And maybe also the file too. 700+ lines for a file is already a bit long.

Feel free to ping me on the split off PRs.

yu-yama committed:
fix: use `rw` tactic instead of `▸`

The former code no longer compiles after updating to v4.16.0-rc1.

yu-yama committed:
docs: improve docstrings

style: use `where` and `__` instead of braces and `with`
… besides one-line definitions
chore: move variables to the top of each namespace

yu-yama committed:
chore: patch lemmas and defs affected by redefinition of `Equiv`

style: move braces to the same lines as fields

yu-yama committed:
refactor: require `smul_eq_conjAct` as a field rather than a lemma

chore: drop `conjActSurjInvRightHom`

yu-yama committed:
feat(GroupExtension/Abelian): define `SemidirectProduct.conjClassesEquivH1`

yu-yama committed:
docs: remove a resolved TODO

YaelDillies committed:
fix

YaelDillies committed:
fix
